### DISCUSSION TITLE
Do certain CIA/VIC2 emulation make sense in the context of this library?

### DISCUSSION THREAD

**Discussion by Toddler-Boy** [libsidplayfp/libsidplayfp#95] [Q&A] [answered]
I know that some code was copied from VICE for CPU, CIA, and VIC2 emulation, and it was probably easier to leave it as is, but I wonder if we need code to handle lightpens, the serial port, potentiometers, or sprites.

It can affect CPU timing, but for regular *.sid files, this shouldn't matter as they do not move sprites around on the (invisible) screen. They don't load any additional files via the serial port, and since we can't connect a lightpen, that interrupt should also never be triggered.

So would it theoretically be possible to remove all that code, save some host CPU time without ill effects?

**Comment by drfiemost** [comment, marked as answer]
Well, I know that some tunes rely on perfect timing to play digis (i.e. /MUSICIANS/M/Mahoney/Kapla_Caves.sid or /MUSICIANS/S/Swallow/Wonderland_XII-Digi_part_1.sid), but I can't tell if lightpen or serial port are really needed. The code was there from the start and I've only made it more accurate. The sprites may affect the badline timing in code ripped from demos so I'm quite positive the emulation, which is actuallypretty basic, is required for some sid files.

**Comment by Toddler-Boy** [comment]
So I figured that at least removing the serial port, potentiometers, and lightpen should be safe. Yes, the code is not too long or complicated, but doing something a million times still adds up ;)

**Comment by Toddler-Boy** [comment]
Can you give me a small list of *.sid files you use for specific tests? I heard the Kapla_Caves demo before, but didn't know its name and that it requires exact timing. It would be nice to have a list of test files I can use for regression tests.

I would render a WAV file at e.g., 48 kHz and then write a small tool that renders them again and verifies that the output hasn't changed after some internal changes/optimizations.

**Reply by drfiemost** [reply]
I'm trying to collect some interesting tunes for testing in the wiki:
https://github.com/libsidplayfp/libsidplayfp/wiki/Interesting-SID-tunes

Thinking about it, with Kapla_Caves I had only some problem with the filter. The Wonderland digi tunes were more demanding and experienced some problems with the VIC timing.

By the way, if you want to compare the output of two runs be aware of the random startup delay.

**Reply by Toddler-Boy** [reply]
> By the way, if you want to compare the output of two runs be aware of the random startup delay.

Thanks. I'm already always using the MAX_POWER_DELAY time, so when measuring output volume for my replay-gain database, it is consistent. I also think it's realistic to assume that when a user turns on his C64 and then proceeds to load a tune and start it, more than 8,000 cycles have passed ;)

I will probably increase this delay even further to ensure that all (simulated) capacitors, pull-downs, etc., have happened before the tune starts.

Some tunes have an unpleasant plop (sounds like volume register change?) sound at the beginning. It would be nice to avoid it somehow. Maybe create a database with a start offset so the first "n" output samples are ignored.

**Comment by drfiemost** [comment]
> Some tunes have an unpleasant plop (sounds like volume register change?) sound at the beginning. It would be nice to avoid it somehow. Maybe create a database with a start offset so the first "n" output samples are ignored.

Yes, that's the volume register, most noticeable on 6581. Ideally one should start producing sound after the system initialization is done, it would help for #82 too, but that's easier said than done.
